# Assignment 8, Part 1: Pandas

This notebook is based on the official `pandas` [documentation](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html).  Unless otherwise credited, quoted text comes from this document.

`Pandas` is a Python package that we will use to manage labeld data.

## Overview

*Where does this tool fit in my life?*

> For data scientists, working with data is typically divided into multiple stages: munging and cleaning data, analyzing / modeling it, then organizing the results of the analysis into a form suitable for plotting or tabular display. pandas is the ideal tool for all of these tasks.


*Why use this package?*

> pandas is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, real world data analysis in Python

### Data Structures

Pandas provides two new objects in which to store and manipulate data.

Dimensions | Name |	Description
-----------|-------|--------------------------------------
1 |	Series | 	1D labeled homogeneously-typed array
2 |	DataFrame | 	General 2D labeled, size-mutable tabular structure with potentially heterogeneously-typed column

Typical usage of this package involves the creation of one or more `DataFrame` objects, and manipulating them with the provided methods.

### Pandas Cheat Sheet

*"[`Package Name`] cheat sheet"* for popular packages can often be a worthwhile internet query.

In the case of pandas, there is an official cheet sheat you can find [here](http://pandas.pydata.org/Pandas_Cheat_Sheet.pdf).


### 10* Minutes to Pandas

The official [10 Minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html#min) documentation is an excellent resource, and provides the titular guide which we will drawn from and expanded upon for this tutorial.

_Note: You can access the pandas documentation from within a jupyter lab session by selecting 'help' and then selecting 'pandas' from the dropdown menu._

---
## 1. Getting Started

First, we need to import the pandas library (and Numpy library too).  All packages are imported at the top of the notebook. Execute the code in the following cell to get started with this notebook (type Ctrl+Enter in the cell below)

In [1]:
# Numpy and pandas usage are often intertwined.
# These abbreviations are ubiquitiously used.
import numpy as np
import pandas as pd

The code above imports both numpy and pandas as variables named `np` and `pd`. We can use these variables to access the functionality of both libraries respectively.  The above is what we will use for the rest of this class.

You may be wondering why we didn't import pandas like this:  
```python
import pandas
```
We could, but the first is far more commonly seen. 

#### Task 1.1: Setup


---
## 2. Data Object Creation

Here we will learn to create and use the two data structures provided by Pandas: **Series** and **DataFrames**.  For additional help, see the [Data Structure Intro section](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dsintro).


### 2.1 The `pd.Series` Data Object

In Pandas, a Series is a one dimensional array, where all the entries are of the same data type (i.e. integer, string, boolean, etc.) or **NaN**.  NaN means "not a number" but serves as a marker when a value is missing.  Before using a series take care to consider the correct data type.

The following cell creates an new series containing an array of 5 numbers and one missing value by simply passing a list of 6 elements to the `pd.Series` function. Execute the cell:

In [ ]:
# Create the series object.
my_series = pd.Series([1, 3, 5, np.nan, 6, 8])

Execute the next cell to see what the new series object named `my_series` contains

In [ ]:
# Call the object in the notebook.
# This (mostly) calls the __repr__() function of a given object.
my_series

#### Task 2a Create a `pd.Series` object
Use the practice notebook to create a series of your own design.

### 2.2 The `pd.DataFrame` Data Object

In Pandas, a DataFrame is a two-dimensional array.  Just like Numpy, it has axes labeled "rows" and "cols".  However, unlike Numpy, it allows for different data types in each column!  DataFrames also provide greater flexibilty for indexing and slicing.

#### 2.2.1 Create a DataFrame with a dictionary

A dataframe is created by passing a Python dictionary (i.e. **dict**) to the `pd.DataFrame` function.  Remember dictionaries have key/value pairs. The dictionary is interpreted in the following way:

- The keys in the dictionary argument becomes the column names
- The value in the dictionary argument contains the column values
- Each "column" must have the same number of values.

The following cell contains code that creates a new DataFrame object named `df_1`.  This dataframe consists of two columns, one named "alpha" and the other "beta".  Each column has 5 elements. Notice that the columns contain different data types.  Execute the following cell.

In [2]:
# Creating a dataframe with a dictionary.
df_1 = pd.DataFrame(
    {'alpha': [0, 1, 2, 3, 4],
     'beta': ['a', 'b', 'c', 'd', 'e']})

Execute the next cell to see what the new data frame object named `df_1` contains

In [3]:
# Call the object in the notebook.
# This time we get a html output, as the notebook shell
# sees a _repr_html_() function to call.
df_1

alpha beta
0      0    a
1      1    b
2      2    c
3      3    d
4      4    e

Observe that a DataFrame can be visualized as a data table or Excel spreadsheet. Also, an index, in the left-most column, has been automatically generated.

#### 2.2.2. Create a DataFrame with a Numpy array
Alternatively, we can create a DataFrame by passing in a Numpy matrix.  Remember that this is only possible if the data is all of the same type as Numpy arrays can only store one data type. The following code creates a Dataframe of 6 rows and 4 columns and fills it with random numbers generated using the Numpy `np.random.randn` function.  Execute the following cell:

In [6]:
df = pd.DataFrame(np.random.randn(6, 4))

Let's take a look at our new data frame by executing the following cell:

In [7]:
df

0         1         2         3
0 -0.531861 -0.864166  0.766840  0.086372
1 -1.489222  0.821623  0.248321  0.845579
2 -1.158525 -0.365787 -1.209259 -0.241529
3  0.112887 -2.128430  0.646099 -0.408736
4 -0.407423  1.808536  0.362305 -2.251657
5 -0.588663  1.019797  1.600386  0.478823

Observe that because we did not provide a dictionary, we do not have header names for each column. Rather a numeric index was automatically generated.

#### Task 2b: Create a pd.DataFrame object from a Python dictionary

### 2.3. Create a DataFrame with row and column labels

Rather than have a numeric index for each row of the DataFrame, we may want to give each row a human-readable label. This will help make the data more accessible (as will be seen later).  

To demonstrate, lets reuse the 6x4 datafame we created in section 2.2.2.  First, we need to provide our index names.  Suppose we want our row names to be dates, perhaps corresponding to when the row data was collected. The code below uses the `pd.date_range` function to do create a list of dates. If this were real data we would most likely include the dates collected as a column in the table, but for the purposes of demonstration we'll generate a series of regularly spaces dates over a 6 day period: one for each of the 6 rows:

In [4]:
dates = pd.date_range('20130101', periods=6)
dates

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06'],
              dtype='datetime64[ns]', freq='D')

Let's also suppose that we have 4 columns and we want them named 'A', 'B', 'C', and 'D'. 

We can re-create the DataFrame from section 2.2.2 with these names for the columns and the dates as row indexes by using the `index` and `columns` arguments when creating the DataFram as shown in the following cell.  Execute the cell to see the results: 

In [8]:
df = pd.DataFrame(np.random.randn(6, 4), index=dates, columns=list('ABCD'))
df

A         B         C         D
2013-01-01  0.025308  0.553918  0.582617  0.317441
2013-01-02 -1.864706 -1.157520 -0.144645  2.062387
2013-01-03  0.739027 -0.982459  0.046915 -1.179101
2013-01-04  0.484775  1.356177 -0.134019  2.225083
2013-01-05 -0.632753 -0.795577 -0.748937 -1.365607
2013-01-06  0.428006  0.684617 -0.745004 -0.256929

#### Task 2c: Supply indexes and columns when creating a pd.DataFrame object.


---
## 3. Loading Data

Creating a DataFrame manually, as in the previous examples, would be tedious for large datasets.  Fortunately, Pandas provides a variety of functions for importing data from files.  This saves you from writing your own Python code to open a file, read in the contents and format it for inclusiong into a DataFrame. For this tutorial we will look at two of these import functions:  `pd.read_csv` and `pd.read_excel()`.

For this portion of the tutorial, we will use a rather [famous set of data concerning iris flowers](https://en.wikipedia.org/wiki/Iris_flower_data_set).  You can find a copy of the file [by clicking here]( https://gist.githubusercontent.com/curran/a08a1080b88344b0c8a7/raw/d546eaee765268bf2f487608c537c05e22e4b221/iris.csv).

![Iris flower](https://upload.wikimedia.org/wikipedia/commons/thumb/4/41/Iris_versicolor_3.jpg/193px-Iris_versicolor_3.jpg)

To use this file, create a directory named `data` in the same directory as this notebook and save the file in that directory with the name `iris.csv`.  

#### Task 3a: Download the iris.csv file.
Download the iris data and take a moment to learn about this data from the Wikipedia page cited above

### 3.1 Importing Using `pd.read_csv`

The `pd.read_csv` function can be used to read in files that are in comma-separated format or tab delimited format: two very common data formats.  This function is very flexible and has a large number of arguments. See the [`pd.read_csv` online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) for a thorough listing of arguments. 

The iris data is in CSV format and is compatible with the default arguments for the `pd.read_csv` function, therefore we can easily import this data by providing the path to the file as the only argument to the function. This will automatically create a DataFrame for us! Execute the following line to import the data:

In [13]:
iris_df = pd.read_csv('data/iris.csv')

#### Task 3b: Import the iris.csv file
In your own practice notebook, import the iris dataset.

#### Task 3c: Import a tab delimited file
Take a look at the `pd.read_csv` online documentation. Describe how you would import a tab-delimited file.

### 3.2 Importing using `pd.read_excel()`
It is common for data to be stored in Excel worksheets (provided the data is not too large).  You can easily export data from Excel to CSV or Tab delimited formats but Pandas provides  the `pd.read_excel` function to save you the time.  We will not practice loading data from Excel, but take a look at the [`read_excel` online documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html) for details.

---
## 4. Explore your Data
There are a variety of functions for exploring data once you have imported it. Here we will discuss `pd.head`, `pd.tail`, `pd.sample`, and `pd.describe`.

### 4.1 The `pd.head` function

As with any Python variable, you can display its contents by simply typing the variable name and pressing the "enter" key.  Howevever, for very large data files this may not be desired.  The data used in this tutorial is not extremly large but this often not the case.

To easily examine your the first set of rows in your DataFrame, use the `head()` function of the `pd.DataFrame` object. By deafult it displays the first 5 rows. Execute the following cells to view the first five rows of the iris dataset.

In [10]:
iris_df.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

This is an excellent way to check if you created your `pd.DataFrame` object correctly.  You can change the number of rows returned by providing an argument. The following requests the first 10 rows of the data frame:

In [12]:
iris_df.head(10)

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa
6           4.6          3.4           1.4          0.3  setosa
7           5.0          3.4           1.5          0.2  setosa
8           4.4          2.9           1.4          0.2  setosa
9           4.9          3.1           1.5          0.1  setosa

### 4.2 The `pd.tail` function
Simliar to the `pd.head` function, the `pd.tail` function lets you peak at the last 5 rows of data.  To demonstrate, execute the following cell.

In [17]:
iris_df.tail(5)

sepal_length  sepal_width  petal_length  petal_width    species
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

### 4.3 The `pd.sample` function
The `pd.sample` function selects a random set of rows to display. By default it only selects 1 row, we can provide the number of rows we would like (e.g. 5) by providing the number as an arugment:

In [18]:
iris_df.sample(5)

sepal_length  sepal_width  petal_length  petal_width     species
90            5.5          2.6           4.4          1.2  versicolor
57            4.9          2.4           3.3          1.0  versicolor
15            5.7          4.4           1.5          0.4      setosa
106           4.9          2.5           4.5          1.7   virginica
7             5.0          3.4           1.5          0.2      setosa

Repeat execution of the function to see a different set of randomly selected rows:

In [19]:
iris_df.sample(5)

sepal_length  sepal_width  petal_length  petal_width    species
1             4.9          3.0           1.4          0.2     setosa
132           6.4          2.8           5.6          2.2  virginica
124           6.7          3.3           5.7          2.1  virginica
46            5.1          3.8           1.6          0.2     setosa
9             4.9          3.1           1.5          0.1     setosa

### 4.4 The `pd.describe` function
The `pd.describe` function helps demonstrate the power of Pandas over Python dictionaries and Numpy arrays.  It provides a summary of statistics for the data frame in one simple function call. These statistics include the count, mean, standard deviation, min, max and quartiles.

Execute the following cell to demonstrate:

In [21]:
iris_df.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

In [23]:
iris_df.corr()

sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000    -0.109369      0.871754     0.817954
sepal_width      -0.109369     1.000000     -0.420516    -0.356544
petal_length      0.871754    -0.420516      1.000000     0.962757
petal_width       0.817954    -0.356544      0.962757     1.000000

In [25]:
iris_df.corr().values

array([[ 1.        , -0.10936925,  0.87175416,  0.81795363],
       [-0.10936925,  1.        , -0.4205161 , -0.35654409],
       [ 0.87175416, -0.4205161 ,  1.        ,  0.9627571 ],
       [ 0.81795363, -0.35654409,  0.9627571 ,  1.        ]])

In [26]:
iris_df.corr('kendall')

sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000    -0.072112      0.717624     0.654960
sepal_width      -0.072112     1.000000     -0.182391    -0.146988
petal_length      0.717624    -0.182391      1.000000     0.803014
petal_width       0.654960    -0.146988      0.803014     1.000000

#### Task 4a: Use head, tail and sample with the iris data frame.


---
## 5. Accessing Data

In this section we will learn to retreive values, slice the data frame, and use column and row labels (rather than numeric indexes) to find data. 

### 5.1 Viewing column and row labels.
Before we delve into retreiving data, lets first look at some properties of data frames for retrieving the dataframe column and row names.  This is because we often want to work with, or check the indexes and columns of a data frame using their names.  First, to retrieve the columns names we can use the `columns` property of the data frame:

In [27]:
iris_df.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

Similary, we can get a description of the index using the `index` property:

In [28]:
iris_df.index

RangeIndex(start=0, stop=150, step=1)

Observe that here our index is numeric and runs from 0 to 150, incremented by 1

#### Task 5a: Display the columns and indexes of the iris data frame.


### 5.2. Get all the values

Perhaps you would like to extract the contents of the DataFrame into a 2-dimensional Python list. You will lose the power of Pandas, but you should know how to do this. There are two ways to extract values: the new Pandas v0.24 way and the older way.  

This following code example used to be the standard way of getting the values of a dataframe as a `numpy` array:

```python
my_array = df.values
```

This is depreciated in version 0.24 of `pandas` to instead require the use of `to_numpy()`.

```python
my_array = df.to_numpy()
```

You can verify what version of `pandas` you have:

In [31]:
pd.__version__

'0.23.4'

In [32]:
vals = df.values
vals

array([[ 0.02530841,  0.55391827,  0.58261702,  0.31744065],
       [-1.8647057 , -1.15752042, -0.14464537,  2.06238703],
       [ 0.73902652, -0.98245856,  0.04691495, -1.17910068],
       [ 0.48477485,  1.356177  , -0.13401943,  2.22508318],
       [-0.63275274, -0.79557711, -0.74893673, -1.36560665],
       [ 0.4280064 ,  0.68461716, -0.74500355, -0.25692924]])

#### Task 5b: Check the version of `pandas` you have, then convert a data frame to a numpy array.



In [ ]:
# iris_df.values

### 5.3 Basic Selections
Pandas provides a variety of row/column selectors that are similar to the way that lists and arrays are indxed in Python and Numpy. This tutorial will introduce those. However, please note:
> While standard Python / Numpy expressions for selecting and setting are intuitive and come in handy for interactive work, for production code, we recommend the optimized pandas data access methods, '.at', '.iat', '.loc' and '.iloc'.

This tutorial will also descirbe the `.at`, `.iat`, `.loc` and `.iloc` functions.

#### 5.3.1 Get a column
The easiest way to select a single column is to use the column label! When in doubt, use the `.column` property of the DataFrame to find the list of column labels.  Using the column label will result in the column being returned as a Series object.  For example, if we want the `sepal_length` column of the iris data frame we can execute the following:

In [ ]:
slength = iris_df['sepal_length']
slength.head()

Observe, we can use the `.head` function to get the first few rows of a Series as well as a DataFrame.

#### 5.3.2 Get rows (slicing)
Similar to slicing of Python lists, we can do the same with Pandas DataFrames:

```python
iris_df[start:end]
```

As a reminder, indexing of a list in Python uses 0-based indexing and negative numbers index in reverse order:

```
 +---+---+---+---+---+---+
 | P | y | t | h | o | n |
 +---+---+---+---+---+---+
   0   1   2   3   4   5  
  -6  -5  -4  -3  -2  -1
```

**Important:** Recall the recommendation to *not* using this type of indexing (slicing) in production with very large data. Instead use other provided functions (`.loc`, `iloc`, etc). However, to practice, let's get the first row of the iris dataset by slicing:

In [33]:
iris_df[0:1]

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa

Next lets get the first five rows:

In [34]:
iris_df[0:5] 

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

### 5.4. Selecting with `df.loc` and `df.iloc`

Aside from the column name and slicing (as shown in the previous sections) Pandas provides the `df.loc` and `df.iloc` accessor.  These accessors perform better for selecting subsets of the dataframe. The `.loc` accessor is used for slicing by columns and `.iloc` is used for slicing by rows.

### 5.4.1 Using `.loc`

The `.loc` accessor allows for the slicing of the data frame by using the labels of either rows and columns.  

The syntax is as follows:

```python
df.loc[row_start:row_end, column_start:column_end]
```

Remember, because the rows in the iris data frame are indexed using integers, we can use `.loc` to get rows using the integer values.  To get the first row:

In [35]:
iris_df.loc[0]

sepal_length       5.1
sepal_width        3.5
petal_length       1.4
petal_width        0.2
species         setosa
Name: 0, dtype: object

To get the first two rows:

In [36]:
iris_df.loc[0:1]  # Get the first two rows.

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

Observe that when only one row is returned that a Series is provided.  When multiple rows are returned then a new DataFrame object is returned.


Next, recall that we constructed a 6x4 dataframe in section 2.3 that was filled with random values and we set the row names using datetime values. The data frame was named `df`.  As a reminder here are the first 2 rows:

In [ ]:
df.head(2)

In this data frame the rows are indexed using strings that represent dates. We can therefore use those labels to slice the dataframe.  The following extracts the first row:

In [ ]:
df.loc['2013-01-01']

Similarly we can get the first two rows:

In [ ]:
df.loc['2013-01-01':'2013-01-02']

#### 5.4.2 Using `.loc` to get columns
Remember for our iris data frame the column indexes have the names as provided by the `.columns` property.

```python
iris_df.columns
Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')
```

We can use these names to retrive values for specific columns. For example, the code below extracts rows 4 through 6 and the columns 'sepal_width' and 'petal_width':

In [ ]:
iris_df.loc[4:6, 'sepal_width': 'petal_width']

The use of the colon in the code above between 'sepal_width' and 'petal_width' implies a range. Therefore every column between those two are included. This is why 'petal_length' appears in the output.  If we wanted to specifically select only the two columns and none other, we could provide the names in a list:

In [ ]:
iris_df.loc[4:6, ['sepal_width', 'petal_width']]

Observe that now we only have 'sepal_width', and 'petal_width' in the resulting data frame.

Moreover, we can limit the rows to only those specified in the same way:

In [ ]:
iris_df.loc[[4,6], ['sepal_width', 'petal_width']]

#### Task 5c: Make selections with `loc`

+ Select a single item with `at`.
+ Select a row slice with `loc`.
+ Select a row and column slice with `loc`.

### 5.4.2 Using `.iloc`

The `.iloc` accessor allows us to slice the data frame using an integer index, regardless of what the index (or column) labels actually are.

The syntax is the same as `.loc` with the exception of that integers are used instead of labels:

```python
df.iloc[row_start:row_end, column_start:column_end]
```
To retrieve the first row of the iris data we use the numeric index:

In [ ]:
iris_df.iloc[0]  # Get the first row.

We can use any combination of integer indexes to get a subset of rows and columns.  The following examples retieves rows 2 to 5, and excludes the 'species' column.

In [ ]:
iris_df.iloc[2:5, :-1] 

#### Task 5d: Make selections with `iloc` and `iat`.

+ Select a single item with `iat`.
+ Select a row slice with `iloc`.
+ Select a row and column slice with `iloc`.


### 5.5 "Fancy" indexing with `.where`

Pandas allows us to use boolean values to extract specific rows of data that meet certain conditions.To clarify this, first examine what occurs when we apply a condition to a data frame.  Recall the data frame we created in section 2.2.1.  As a reminder execute the following cell:

In [37]:
df_1

alpha beta
0      0    a
1      1    b
2      2    c
3      3    d
4      4    e

In [38]:
df_1[df_1 > 1.0]

alpha beta
0    NaN    a
1    NaN    b
2    2.0    c
3    3.0    d
4    4.0    e

In [39]:
df_1 > 1.0

alpha  beta
0  False  True
1  False  True
2   True  True
3   True  True
4   True  True

Let's apply a condition to the data frame to find all values > 1.0

In [40]:
df_1 > 1.0

alpha  beta
0  False  True
1  False  True
2   True  True
3   True  True
4   True  True

Notice that the resulting data frame contains `True` and `False` values in place of every value that did or did not meet the condition.  We can use this new data frame to subset the original dataframe.

Lets consider the iris dataset for a more realistic example.  Suppose we want to find all rows with a `sepal_length` greater than 5.8. We know from the call to `describe` that the mean is approximately 5.8. We can do so with the following:

In [ ]:
condition = iris_df['sepal_length'] > 5.8
iris_df[condition].head()

You could rewrite the two lines above to combine them into a single line with identical results:

In [ ]:
iris_df[iris_df['sepal_length'] > 5.8].head()

#### Task 5e: Boolean Indexing

Create subsets using boolean indexes that:
+ Use one boolean operator.
+ Use two boolean operators.
